File: packages/torqued-graph/src/json.ts
/**
 * JSON (de)serialisation helpers for fields stored as `String` in the
 * Prisma schema (a necessity on SQLite — see `prisma/schema.prisma` notes).
 *
 * Use `encodeJson` on writes and `decodeJson<T>` on reads. Both helpers
 * tolerate `null` / `undefined` and round-trip cleanly with the Python
 * mirror in `packages/torqued-graph/python/kinform_torqued_graph/json.py`.
 */

export function encodeJson(value: unknown): string {
  return JSON.stringify(value ?? null);
}

export function encodeJsonNullable(value: unknown | null | undefined): string | null {
  if (value === null || value === undefined) return null;
  return JSON.stringify(value);
}

export function decodeJson<T>(raw: string): T {
  return JSON.parse(raw) as T;
}

export function decodeJsonNullable<T>(raw: string | null | undefined): T | null {
  if (raw === null || raw === undefined) return null;
  return JSON.parse(raw) as T;
}

File: packages/torqued-graph/python/kinform_torqued_graph/json.py
"""JSON helpers for fields stored as TEXT in the Prisma schema.

Mirror of ``packages/torqued-graph/src/json.ts``. Use :func:`encode_json` on
writes and :func:`decode_json` on reads.
"""

from __future__ import annotations

import json
from typing import Any, TypeVar

T = TypeVar("T")


def encode_json(value: Any) -> str:
    return json.dumps(value if value is not None else None, separators=(",", ":"))


def encode_json_nullable(value: Any) -> str | None:
    if value is None:
        return None
    return json.dumps(value, separators=(",", ":"))


def decode_json(raw: str) -> Any:
    return json.loads(raw)


def decode_json_nullable(raw: str | None) -> Any:
    if raw is None:
        return None
    return json.loads(raw)
